dp/generate_climos.py

```diff
@@ -39,6 +39,86 @@
 cdo = Cdo()
 
 
+def has_climo_periods(input_file, climo):
+    '''
+    This function checks if the given input NetCDF dataset has a variable "climo_periods".
+    generate_climos requires the input dataset to have the variable, thus this function will
+    warn if there is no such variable.
+    '''
+    periods = input_file.climo_periods.keys() & climo
+    logger.info("climo_periods: {}".format(periods))
+    if len(periods) == 0:
+        logger.info(f"{input_file.filepath()} has no variable 'climo_periods'")
+    return periods
+
+
+def is_NetCDF(filepath):
+    '''
+    This function checks if the given input file path is a vaild NetCDF file. It returns
+    NetCDF dataset of the input file. Ohterwise, it halts the program.
+    '''
+    try:
+        return CFDataset(filepath)
+    except Exception as e:
+        logger.info("{}: {}".format(e.__class__.__name__, e))
+        sys.exit()
+
+
+def dry_run_handler(filepath, climo):
+    '''
+    This function handles dry-run operation of generate_climos. The purpose of this function
+    is to check variables and attrbutes to be used for generate_climos. dry-run will not 
+    produca any output files.
+    '''
+    logger.info("")
+    logger.info("File: {}".format(filepath))
+    input_file = is_NetCDF(filepath)
+    has_climo_periods(input_file, climo)
+
+    for attr in ['project', 'institution', 'model', 'emissions', 'run']:
+        try:
+            logger.info(
+                "{}: {}".format(attr, getattr(input_file.metadata, attr))
+            )
+        except Exception as e:
+            logger.info("{}: {}: {}".format(attr, e.__class__.__name__, e))
+    logger.info(
+        "dependent_varnames: {}".format(input_file.dependent_varnames())
+    )
+    for attr in ['time_resolution', 'is_multi_year_mean']:
+        logger.info("{}: {}".format(attr, getattr(input_file, attr)))
+
+
+def generate_climos(
+    filepath,
+    outdir,
+    operation,
+    **kwargs,
+):
+    '''
+    This function runs general generate_climos operation. The main purpose of this function
+    is to call create_climo_files.
+    '''
+    logger.info("")
+    logger.info("Processing: {}".format(filepath))
+    input_file = is_NetCDF(filepath)
+    periods =  has_climo_periods(input_file, kwargs["climo"])
+
+    for period in periods:
+        t_range = input_file.climo_periods[period]
+        create_climo_files(
+            period,
+            outdir,
+            input_file,
+            operation,
+            *t_range,
+            convert_longitudes=kwargs["convert_longitudes"],
+            split_vars=kwargs["split_vars"],
+            split_intervals=kwargs["split_intervals"],
+            output_resolutions=kwargs["resolutions"]
```

Comment:
What is the behaviour of these variables if the keys you are searching for are not available?

Reply:
Default values for stdin arguments in [generate_climos](https://github.com/pacificclimate/climate-explorer-data-prep/blob/master/scripts/generate_climos#L60) script assure that keys are always available. We might also want to define default arguments in `thunderbird` as well.

Reply:
Definitely don't recommend `kwargs` approach. See my comments below.